snorkel/contrib/deps/labeling/model/dep_learner.py

```diff
@@ -0,0 +1,192 @@
+                    deps_singleton.append((i, k))
+                if (j == k) and (i < l):
+                    deps_singleton.append((i, l))
+                if (i == l) and (j < k):
+                    deps_singleton.append((j, k))
+        all_deps = list(set(deps_singleton))
+
+        if len(all_deps) == sp.special.comb(M, 2):
+            raise ValueError(
+                "Dependency structure is fully connected. Rerun with higher thresh_mult."
+            )
+        return all_deps
+
+    def _get_deps_from_inverse_sig(
+        self, J: np.ndarray, thresh: float
+    ) -> List[Tuple[int, int]]:
+        """Select values larger than thresh in J as dependent LF indices."""
+        deps = []
+        for i in range(J.shape[0]):
+            for j in range(J.shape[1]):
+                if abs(J[i, j]) > thresh:
+                    deps.append((i, j))
+        return deps
+
+    def fit(
+        self,
+        L: np.ndarray,
+        thresh_mult: Optional[float] = 0.5,
+        gamma: Optional[float] = 1e-8,
+        lam: Optional[float] = 0.1,
+        verbose: Optional[bool] = False,
+    ) -> List[Tuple[int, int]]:
+        r"""Learn dependencies among LFs.
+
+        Learn dependencies using robust PCA based method. In case of multi-class, the method considers log2(cardinality) splits to calculate agreements and disagreements. Solves the following problem
+
+        (\hat{S}, \hat{L}) = argmin \mathcal{L}(S - L, \Sigma_O) + lam*(gamma*||S||_1 + ||L||_*)
+
+        s.t. S - L \succ 0, L \succeq 0
+
+        Note: Best performance requires hyperparameter search over the thresh parameter.
+
+        Parameters
+        ----------
+        L
+            An [n,m] matrix with values in {-1,0,1,...,k-1}
+        thresh_mult
+            Threshold multiplier for selecting thresh_mult * max off diagonal entry from sparse matrix
+        gamma
+            Parameter in objective function related to sparsity
+        lam
+            Parameter in objective function related to sparsity and low rank
+        verbose
+            Whether solver is verbose
+
+        Raises
+        ------
+        ValueError
+            If L has higher cardinality than passed in or solver fails
+
+        Returns
+        -------
+        List[Tuple[int, int]]
+            List of tuples denoting dependencies among LFs
+        """
+
+        N = float(np.shape(L)[0])
+        M = np.shape(L)[1]
+        O_all = np.zeros((M, M))
+
+        if L.max() + 1 > self.cardinality:
+            raise ValueError(
+                f"Does not match DependencyLearner cardinality={self.cardinality}, L has cardinality {L.max()+1}"
+            )
+
+        # calculate agreement-disagreement rates based on random class splits
+        L_shift = np.copy(L)
+        split_list = np.random.choice(
+            range(1, self.cardinality),
+            size=np.int(np.log2(self.cardinality)),
+            replace=False,
+        )
+        for class_thresh in split_list:
```

Comment:
Here, the splits will almost always put adjacent classes in the same group. For a random split, you can do np.random.choice([-1,1],[cardinality, log(cardinality)]), and use each row of that as a split.

Reply:
the problem with that for small number of classes is there's no guarantee the classes will end up in different splits